org.eclipse.buildship.ui.test/src/main/groovy/org/eclipse/buildship/ui/SWTBotTestHelper.groovy

```diff
@@ -0,0 +1,94 @@
+package org.eclipse.buildship.ui
+
+import org.eclipse.swtbot.eclipse.finder.SWTWorkbenchBot
+import org.eclipse.swtbot.eclipse.finder.widgets.SWTBotView
+import org.eclipse.swtbot.swt.finder.exceptions.WidgetNotFoundException
+import org.eclipse.swtbot.swt.finder.widgets.SWTBotShell
+import org.eclipse.swtbot.swt.finder.finders.UIThreadRunnable
+import org.eclipse.swtbot.swt.finder.results.VoidResult
+import org.eclipse.swtbot.swt.finder.results.BoolResult
+import org.eclipse.ui.PlatformUI
+import org.eclipse.core.runtime.jobs.Job
+import org.eclipse.swt.widgets.Display
+
+class SWTBotTestHelper {
+    static SWTWorkbenchBot swtBot;
+
+    public static SWTWorkbenchBot getBot() {
+        if (swtBot == null) {
+            return new SWTWorkbenchBot()
+        }
+        return swtBot
+    }
+
+    private static void closeWelcomePageIfAny() throws Exception {
+        try {
+            SWTBotView view = getBot().activeView()
+            if (view.getTitle().equals("Welcome")) {
+                view.close()
+            }
+        } catch (WidgetNotFoundException e) {
+            UiPlugin.logger().error("Failed to initialize SWTBot test.", e)
+        }
+    }
+
+    public static void closeAllShellsExceptTheApplicationShellAndForceShellActivation() {
+        SWTWorkbenchBot bot = getBot()
+
+        // in case a UI test fails some shells might not be closed properly, therefore we close
+        // these here and log it
+        SWTBotShell[] shells = bot.shells()
+        for (SWTBotShell swtBotShell : shells) {
+            if (swtBotShell.isOpen() && !isEclipseApplicationShell(swtBotShell)) {
+                bot.captureScreenshot(swtBotShell.getText() + " NotClosed.jpg")
+                UiPlugin.logger().warn(swtBotShell.getText() + " was not closed properly.")
+                swtBotShell.close()
+            }
+        }
+
+        // http://wiki.eclipse.org/SWTBot/Troubleshooting#No_active_Shell_when_running_SWTBot_tests_in_Xvfb
+        UIThreadRunnable.syncExec(new VoidResult() {
+
+                    @Override
+                    public void run() {
+                        PlatformUI.getWorkbench().getActiveWorkbenchWindow().getShell().forceActive()
+                    }
+                })
+    }
+
+    public static boolean isEclipseApplicationShell(final SWTBotShell swtBotShell) {
+        return UIThreadRunnable.syncExec(new BoolResult() {
+
+                    @Override
+                    public Boolean run() {
+                        return PlatformUI.getWorkbench().getActiveWorkbenchWindow().getShell().equals(swtBotShell.widget)
+                    }
+                })
+    }
+
+    protected static void waitForJobsToFinish() {
+        while (!Job.getJobManager().isIdle()) {
+            delay(500)
```

Comment:
These seem to be magic numbers to me, and I only used 500 because I'm directly copying existing Buildship code. What's the reason behind decreasing it?